daemon/policy.go

```diff
@@ -215,6 +218,7 @@ func (d *Daemon) PolicyAdd(rules policyAPI.Rules, opts *AddOptions) (newRev uint
 			err2 := fmt.Errorf("Unable to recompile base programs: %s", err)
 			log.WithError(err2).WithField("prefixes", prefixes).Warn(
 				"Failed to recompile base programs due to prefix length count change")
+			d.policy.Mutex.Unlock()
```

Comment:
Do we really need to have policy lock while compiling a bpf program?

Reply:
Because FQDN policy generates CIDR rules (which have prefix lengths), then yes, we will have to hold the lock. However, I believe that this would only occur upon the first CIDR rule which contains a /32 which didn't before (@joestringer can you confirm?). We could, in `v1.4` always keep a `/32` prefix length around to avoid this issue. FWIW, this need to compile base programs will only occur for kernels >= 4.9 <= 4.11.

Reply:
We already keep a `/32` around all the time anyway since we're guaranteed to need to refer to `/32s`:

https://github.com/cilium/cilium/blob/dc8a343a260594a1a95ff5ede63cd13f8c97f81b/daemon/daemon.go#L846

Reply:
In other words, yes, we need it, but also this should never occur for FQDN, only when a CIDR with a new unique prefix length is added.

Reply:
But coming back to @aanm's question, I'm not sure I fully understand why recompilation of base programs needs to be done while holding the policy lock. This affects ipcache lookups for the base programs like the ones attached to the `cilium_host` device, not endpoint devices (which would be handled through individual endpoint BPF program compilation).

Reply:
Ideally we don't compile the base programs and then later on find out that we're not going to apply the policy and need to recompile the base programs again to back out the policy..

Reply:
Great, having '/32' always around solves the issue , thanks for confirming Joe!